Issue #14889: Create and Document Check Update Procedure for New Lang… (#14904)
checkstyle/checkstyle: docs/NEW_LANGUAGE_FEATURE_INTEGRATION_PROCESS.md

# New Language Feature Check Integration Process [WIP]

## Check Update Procedure

### Consider Similar Tokens

Identify existing tokens that are similar to the new language feature tokens.
This comparison helps to determine which existing checks may be relevant
and require updates to support the new token.

**Examples**

> - for `RECORD_DEF`, it was reasonable to look at all checks

Reviewer: Please do the same for the next example.

that had `CLASS_DEF` in their acceptable tokens.
- for `TEXT_BLOCK_CONTENT`, it was reasonable to look at all checks
that had `STRING_LITERAL` in their acceptable tokens.

### Java Enhancement Proposals (JEPS)

Review Java Enhancement Proposals (JEPS) related to the new feature.
JEPS provide detailed information about the goals
and motivations behind new language features.
This information can help us understand the feature better and figure out
which checks are most likely to be impacted.

**Example**
- Unnamed variables (`_`), JEP guides us to recognize the new role of the
underscore for unnamed variables and avoid flagging the non-use of such variables.
Consequently, we should update `UnusedLocalVariableCheck` to ensure it does not

> incorrectly violate underscore variables.

Reviewer: ```suggestion
incorrectly violate unnamed variables.
```

Consistent language is very important here.

### Always Impacted Checks

Certain checks are most likely to be impacted by new language features.
It is mandatory to analyze impact on the following checks:

- [IllegalToken](https://checkstyle.org/checks/coding/illegaltoken.html)
- [IllegalTokenText](https://checkstyle.org/checks/coding/illegaltokentext.html)
- [Indentation](https://checkstyle.org/checks/misc/indentation.html#Indentation)
- [Whitespace](https://checkstyle.org/checks/whitespace/index.html)

### Review Other Static Analysis Tools

Examining how other static analysis tools handle new language features
can provide valuable insights into best practices and potential pitfalls.

**Example**

> - Suppose Java 17 or higher, IntelliJ IDEA introduced an inspection rule to detect

Reviewer: ```suggestion
- IntelliJ IDEA introduced an inspection rule for Java 17 or higher to detect
```

> redundant `strictfp` modifiers. In this case, we should consider updating `RedundantModifierCheck`
> to have a similar behaviour.

Reviewer: ```suggestion
redundant `strictfp` modifiers. This resulted in analysis of `RedundantModifierCheck`
to verify if we needed to update it.
```

### Real Usages Examples in Large Projects

Review real usage examples of the new language feature in large projects.

> This helps to identify potential issues and ensures that the checks

Reviewer: ```suggestion
This helps to identify potential issues, and ensures that the checks
```

are aligned with practical use cases.
List of representative projects can be found
[here](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/github-action-projects1.properties)

## New Check Procedure

### Good Source of Best Practices

Identify a reliable source of best practices related to the new language feature.
This could include official language documentation, community-driven style guides.
We may consider leveraging the OpenJDK mailing list as a valuable resource
for gathering conventions and guidelines related to the new language feature.

### Review Associated JEP's Recommendations

Refer to the associated JEPs for the new language feature.
JEPs provide detailed insights into the design decisions,
motivations, and intended usage of the feature.
Often times, JEPs may solve a particular problem in the language,
and we can create checks to suggest to use these new features
Pay close attention to any recommendations provided in the JEPs,
as they can inform the development of effective new checks

**Example**

- Pattern Matching for instanceof, where the newer pattern matching feature provides

> a more concise and type-safe alternative to traditional casting. We can create check
> to suggest to use this new feature

Reviewer: ```suggestion
a more concise and type-safe alternative to traditional casting. Ideally, we would consider creating a check
to suggest to use this new feature if we see the typical `if X instanceof Y`, then typecasting pattern. 
```

### Discover Similar Checks

Look for existing checks that address concepts
similar to those introduced by the new language feature.
These checks serve as references for designing and implementing new checks.

**Examples**
- **Identifiers and Naming Conventions**: If the new feature introduces

> new tokens involving identifiers explore existing checks related to naming conventions.

Reviewer: ```suggestion
new tokens involving identifiers, explore existing checks related to naming conventions.
```

> # New Language Feature Check Integration Process [WIP]

Reviewer: Add 1-2 sentences here about what this document is about.

> ## Check Update Procedure

Reviewer: Add 1-2 sentences here about how updating checks to use modern paradigms and support new language features help us to stay relevant.

It is mandatory to analyze impact on the following checks:

- [IllegalToken](https://checkstyle.org/checks/coding/illegaltoken.html)
- [IllegalTokenText](https://checkstyle.org/checks/coding/illegaltokentext.html)
- [Indentation](https://checkstyle.org/checks/misc/indentation.html#Indentation)
- [Whitespace](https://checkstyle.org/checks/whitespace/index.html)

### Review Other Static Analysis Tools

Examining how other static analysis tools handle new language features
can provide valuable insights into best practices and potential pitfalls.

**Examples**:

- IntelliJ IDEA introduced an inspection rule for Java 17 or higher to detect
  redundant `strictfp` modifiers. This resulted in analysis of `RedundantModifierCheck`
  to verify if we needed to update it.

### Real Usages Examples in Large Projects

Review real usage examples of the new language feature in large projects.
This helps to identify potential issues, and ensures that the checks
are aligned with practical use cases.
List of representative projects can be found
[here](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/github-action-projects1.properties)

> ## New Check Procedure

Reviewer: Add 1-2 sentences here about how creating new checks for new language features helps us to establish best practices within the community.

### Good Source of Best Practices

> Identify a reliable source of best practices related to the new language feature.

Reviewer: Let’s find a couple of sources now. We can consider Baeldung, Stackoverflow, and try to find one more that supports your style guide claim.

Author: https://blogs.oracle.com/javamagazine/post/java-19-gems-jeps
https://blog.jetbrains.com/idea/2024/03/java-22-and-intellij-idea/
Oracle and JetBrains blogs are very good

Reviewer: Ah yes, great idea!

This could include official language documentation, community-driven style guides.

> We may consider leveraging the OpenJDK mailing list as a valuable resource

Reviewer: Let’s remove mention of mailing list completely until we hear back from them, and create a ticket so we don’t forget about this.

JEPs provide detailed insights into the design decisions,
motivations, and intended usage of the feature.
Often times, JEPs may solve a particular problem in the language,
and we can create checks to suggest to use these new features
Pay close attention to any recommendations provided in the JEPs,
as they can inform the development of effective new checks

**Examples**:

- Pattern Matching for instanceof, where the newer pattern matching feature provides
  a more concise and type-safe alternative to traditional casting.
  Ideally, we would consider creating a check
  to suggest to use this new feature if we see the typical `if X instanceof Y`,
  then typecasting pattern.

### Discover Similar Checks

Look for existing checks that address concepts
similar to those introduced by the new language feature.
These checks serve as references for designing and implementing new checks.

**Examples**:

- **Identifiers and Naming Conventions**: If the new feature introduces
  new tokens involving identifiers, explore existing checks related to naming conventions.
  We can create a new similar check to enforce naming conventions for the new feature.

> ## How to Create Tracker Issue

Reviewer: Add 1-2 sentences here that describe how the tracker ticket should contain the results of the analysis that we have suggested in this document, in similar sections, and be created for each new language feature to show that we have done our due diligence.

### Review Other Static Analysis Tools

Examining how other static analysis tools handle new language features
can provide valuable insights into best practices and potential pitfalls.

**Examples**:

- IntelliJ IDEA introduced an inspection rule for Java 17 or higher to detect
  redundant `strictfp` modifiers. This resulted in analysis of `RedundantModifierCheck`
  to verify if we needed to update it.

### Real Usages Examples in Large Projects

Review real usage examples of the new language feature in large projects.
This helps to identify potential issues, and ensures that the checks
are aligned with practical use cases.
List of representative projects can be found
[here](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/github-action-projects1.properties)

## New Check Procedure

Creating new checks for new language features helps us establish
best practices within the community, ensuring that developers
utilize these features effectively.

### Good Source of Best Practices

Identify a reliable source of best practices related to the new language feature.
This could include official language documentation, community-driven style guides

> such as stackoverflow, jetBrains blogs and Oracle Java Magazine.

Reviewer: Let’s make this a list with links

Author: done, had to extend the whitelist the spelling job validates links

> # New Language Feature Check Integration Process [WIP]

Reviewer 2: @nrmancuso Do we consider module-info a new language feature? Using that as the worst example, it requires a whole new grammar parser and tree. Wasn't sure if this was suppose to cover things like that or we are just assuming when we come to this document, the grammar changes are done.

Author: > Wasn't sure if this was suppose to cover things like that or we are just assuming when we come to this document, the grammar changes are done.

Yes, we are assuming that the new syntax is already parsable by checkstyle. IMO, updating grammar and tree should be in a separate doc and we already have an issue for this in the project I hope @nrmancuso gets it done😄

Reviewer 2: Then probably just make it straightforward and say this in the document so it is clear.

Reviewer: Yeah, we can just extend the description here to say something like “after adding support for parsing…”

Reviewer: >  Do we consider module-info a new language feature?

I am not sure, I would say that we can check the language feature by release openjdk page and see if it is there. Really, this is “bigger” than a language feature IMO.

Author: > Then probably just make it straightforward and say this in the document

done

This document outlines the procedures for integrating new language features into Checkstyle.
It includes guidelines for how to discover new checks and
existing checks that need updates to ensure that Checkstyle remains
aligned with language enhancements.

> ## Check Update Procedure

Reviewer 2: We will probably expand this section later, but I am not really seeing a procedure here, just more "think abouts" and "reads" this.

I think I will look forward to us using this somewhere and then have this document reference it as a "good" example and things will grow from there. Some of this is just my thoughts when reading this.

`Identify existing tokens` doesn't say how to go about this identification, it seems like it is assumed we are talking about tokens at the same level. I remember seeing something in other issues where we did token scanning and finding checks that had specific tokens.

I have another section on JEPs.

I would love for some sections to contain more specifics like some basic tools for `Other Static Analysis Tools`.

Reviewer: >Identify existing tokens doesn't say how to go about this identification, it seems like it is assumed we are talking about tokens at the same level. I remember seeing something in other issues where we did token scanning and finding checks that had specific tokens.

I think this can stay sort of abstract, and we can refine later if we need to. I think that the RECORD_DEF example does a good job of illustrating what we mean here.

Reviewer: >I would love for some sections to contain more specifics like some basic tools for Other Static Analysis Tools.

Yes, let's make a list with links to a few of the popular ones, like Intellij Inspections, PMD, and Sonarqube.

Author: done

Updating check modules to support new language features aligns Checkstyle
with current best practices in the Java community
and contributes to the project's ongoing evolution.

### Consider Similar Tokens

Identify existing tokens that are similar to the new language feature tokens.
This comparison helps to determine which existing checks may be relevant
and require updates to support the new token.

**Examples**:

- When the `RECORD_DEF` token was introduced, it was reasonable to look at all checks

>   that had `CLASS_DEF` in their acceptable tokens.

Reviewer 2: I don't think it will always be about similar tokens at the same level. We should also consider how existing parents need to be examined for their new children.

For example, suppose we have to add a new expression. We have had discussions before about breaking apart checks for their expression only counterparts. Right curly was one example, and we just had one recently with padding. We should examine those too, regardless if there is really any other similar token or not.

Reviewer: @mahfouz72 please do
 - update this section to be "Consider Similar and Related Tokens"
 - mention that we should consider the impact that a new child token may have
 - add the example of switch expressions and innerassignmentcheck or some other switch expression example

Author: done

- When `TEXT_BLOCK_CONTENT` token was introduced, it was reasonable to look at all checks
  that had `STRING_LITERAL` in their acceptable tokens.

> ### Java Enhancement Proposals (JEPS)

Reviewer 2: Probably `JEPS` should be first so the user understands what the new feature is before diving into examining existing checks.

Then looking at `Review Associated JEP's Recommendations`, it seems like we are splitting apart the same information.

Going with above, I think we should have a common review the JEPs and understand the change better, then updating and creating new checks can talk about their individual specifics that the global started their thinking process.

Reviewer: I agree, but I think we can document/create an issue for this part and push it back

Reviewer 2: #14923

Review Java Enhancement Proposals (JEPS) related to the new feature.
JEPS provide detailed information about the goals
and motivations behind new language features.
This information can help us understand the feature better and figure out
which checks are most likely to be impacted.

**Examples**:

- Unnamed variables (`_`), JEP guides us to recognize the new role of the
  underscore for unnamed variables and avoid flagging the non-use of such variables.
  Consequently, we should update `UnusedLocalVariableCheck` to ensure it does not
  incorrectly violate unnamed variables.

### Always Impacted Checks

Certain checks are most likely to be impacted by new language features.
It is mandatory to analyze impact on the following checks:

- [IllegalToken](https://checkstyle.org/checks/coding/illegaltoken.html)

> - [IllegalTokenText](https://checkstyle.org/checks/coding/illegaltokentext.html)

Reviewer 2: [IllegalTokenText](https://checkstyle.org/checks/coding/illegaltokentext.html) will not always be impacted. It has to have an identifier in it that is customizable by the user. The others seem correct.

https://checkstyle.sourceforge.io/version/7.0/config_coding.html#IllegalTokenText
For example, at one time this check allowed all tokens and it was later deemed a mistake and changed.

Reviewer: I think `most likely` makes this one OK to have here.

Reviewer 2: I am ok to keep things like this, but I think you made it weirder for me. "Most likely" followed by "mandatory". Which is it really? My comment was only on the mandatory portion which precedes the check list and is made as part of the sentence by `:`.

Reviewer: 😂 good point! @mahfouz72 I gave some bad advice for “mandatory” wording here. Let’s make this more like a suggestion for frequently impacted checks.

JEPs provide detailed insights into the design decisions,
motivations, and intended usage of the feature.
Often times, JEPs may solve a particular problem in the language,
and we can create checks to suggest to use these new features
Pay close attention to any recommendations provided in the JEPs,
as they can inform the development of effective new checks

**Examples**:

- Pattern Matching for instanceof, where the newer pattern matching feature provides
  a more concise and type-safe alternative to traditional casting.
  Ideally, we would consider creating a check
  to suggest to use this new feature if we see the typical `if X instanceof Y`,
  then typecasting pattern.

### Discover Similar Checks

Look for existing checks that address concepts
similar to those introduced by the new language feature.
These checks serve as references for designing and implementing new checks.

**Examples**:

- **Identifiers and Naming Conventions**: If the new feature introduces
  new tokens involving identifiers, explore existing checks related to naming conventions.
  We can create a new similar check to enforce naming conventions for the new feature.

> ## How to Create Tracker Issue

Reviewer 2: I am not sure I can describe it, but I find this section and `Share Results` a bit odd. The whole point of creating an issue is to share the results, I was expecting this section to explain specifics of what should be in the issue from the title. You could rewrite the first sentence here to be `The tracker issue should document and share the results `...

Reviewer: @mahfouz72 please do this ^

Author: done

The tracker issue should document the results of the analysis
conducted according to the outlined procedures in this document.
It should include sections for each step of the analysis with the result
of all of our actions. These tracker issues should be created for each
new language feature, demonstrating our due diligence
in integrating the feature into Checkstyle.

### Document Discovery Methods

After thoroughly researching the impact of the new language feature using the previously
mentioned method, document all relevant findings, including impacted checks,
possible static analysis coverage (new Checks), and best practices.

### Share Results

After completing the discovery process for each new language feature,
open tracker issues to share the results obtained from all methods this includes
checks might need to be updated or new Checks

## How to Open Child Issues

> Open child issues for each check that needs to be updated or for new Check related

Reviewer 2: @nrmancuso Is approval needed for the main issue before children are updated? GSoC may be different, but I would hate for someone to open multiple issues and it turns out there is a disagreement from something and multiple issues are found to not be needed anymore.

Reviewer 2: @nrmancuso If features are still in the preview state, we won't be updating our antlr or reviewing this document, correct? I believe we said we won't support preview things until they are finalized.

Are we going to document these previews and find some way to come back to them at a later time?

> Pattern matching for instanceof
This was a preview in Java 14, finalized in Java 15.

https://docs.oracle.com/javase/specs/index.html
I am not seeing anywhere it says what "older" previews have now been finalized in current versions.

Author: > Are we going to document these previews and find some way to come back to them at a later time?

IMO, the best way to do this is by opening issues for new features once they are released as a preview with antlr and approved labels to remember to come back to them later when they are finalized

Reviewer: As far as preview features go, I usually wait until the community demands support and there is at least two previews before updating our grammar.

Reviewer: > Is approval needed for the main issue before children are updated?

Yes, main tracker issue should be approved before we create child tickets.

Reviewer 2: @nrmancuso It is ok to specify this in the document?

Reviewer: Sure!

Author: I am not sure why to mention that. don't we have a rule somewhere that any issue should be approved first before working on it

the same should apply for tracker issues you shouldn't start working on anything related to it before approved

Reviewer: > I am not sure why to mention that. don't we have a rule somewhere that any issue should be approved first before working on it
> 
> the same should apply for tracker issues you shouldn't start working on anything related to it before approved

Good point, I am good either way

Author: @rnveach what do you think

Reviewer 2: We can revisit this later, but we are talking about approving the tracker issue to work on the child issue. Our rules were to get approval before working on a PR.

# New Language Feature Check Integration Process [WIP]

Adding the ability to parse new language features is the beginning of what we need to do.
after that, we need to update existing checks and create new checks

> to support the new language features. This document outlines the procedures

Reviewer 2: `after` should have a capital A since it is the start of a sentence.

The first sentence sounds odd. If this is from https://github.com/checkstyle/checkstyle/pull/14904#discussion_r1623209059 , then we need to be more explicit.
`Adding the ability to parse new language features is the beginning of what we need to do.` => `The ability to parse new language features must be implemented and merged before beginning this document.`

Author: done

Reviewer: “beginning” -> “following”

Reviewer: @mahfouz72 ^

Author: done

> This document outlines the procedures for integrating new language features into Checkstyle.

Reviewer: ```suggestion
The ability to parse new language features must be implemented and
merged before beginning this document. This document outlines the procedures for integrating new language features into Checkstyle, by updating existing checks and creating new ones.
```

Reviewer: This was kind of wordy and redundant.

> It includes guidelines for how to discover new checks and
> existing checks that need updates to ensure that Checkstyle remains
> aligned with language enhancements.

Reviewer: ```suggestion
```

This ended up being a bit repetitive.

## Check Update Procedure

Updating check modules to support new language features aligns Checkstyle
with current best practices in the Java community
and contributes to the project's ongoing evolution.

### Consider Similar and Related Tokens

Identify existing tokens that are similar to the new language feature tokens.
This comparison helps to determine which existing checks may be relevant
and require updates to support the new token. Additionally,
we should consider the impact that a new child token may have
on existing checks.

**Examples**:

- When the `RECORD_DEF` token was introduced, it was reasonable to look at all checks
  that had `CLASS_DEF` in their acceptable tokens.

- When the `TEXT_BLOCK_CONTENT` token was introduced, it was reasonable to look at all checks
  that had `STRING_LITERAL` in their acceptable tokens.

- When switch expressions were introduced, it was reasonable to look at all checks
  that deals with expressions because `LITERAL_SWITCH` may now fall under `EXPR` token.
  This requires analyzing the impact of this new child token on existing checks,
  such as `InnerAssignmentCheck`

> ### Java Enhancement Proposals (JEPS)

Reviewer: Let's make this one first, as @rnveach already alluded to, we should be reading JEPs before anything else.

Reviewer 2: @nrmancuso Are you saying we don't need the issue anymore? https://github.com/checkstyle/checkstyle/issues/14923

Reviewer: Let’s keep the issue and revisit it later in the summer.

Review Java Enhancement Proposals (JEPS) related to the new feature.
JEPS provide detailed information about the goals
and motivations behind new language features.
This information can help us understand the feature better and figure out
which checks are most likely to be impacted.

**Examples**:

> - Unnamed variables (`_`), JEP guides us to recognize the new role of the

Reviewer: ```suggestion
- for unnamed variables (`_`), the JEP guides us to recognize the new role of the
```

  underscore for unnamed variables and avoid flagging the non-use of such variables.
  Consequently, we should update `UnusedLocalVariableCheck` to ensure it does not
  incorrectly violate unnamed variables.

> ### Always Impacted Checks

Reviewer: ```suggestion
### Frequently Impacted Checks
```

I suppose that nothing is "always", as @rnveach pointed out :)

Reviewer 2: > I suppose that nothing is "always"

That is not true, and sorry if I lead you to believe this.

We should always be using this document with new grammar changes. We just very likely won't update every check because of the new grammar.

My main complaint was when you started mixing likely (synonyms for not always) with mandatory (synonym for always) and it didn't really make sense what was trying to be conveyed. Was it really mandatory or was it just a suggestion? A great movie once said "60% of the time, it works every time.".

Reviewer 2: @mahfouz72 I am still good to change this. Please proceed.

Author: Yes done it is frequently now

- [IllegalToken](https://checkstyle.org/checks/coding/illegaltoken.html)
- [IllegalTokenText](https://checkstyle.org/checks/coding/illegaltokentext.html)
- [Indentation](https://checkstyle.org/checks/misc/indentation.html#Indentation)
- [Whitespace](https://checkstyle.org/checks/whitespace/index.html)

### Review Other Static Analysis Tools

Examining how other static analysis tools handle new language features
can provide valuable insights into best practices and potential pitfalls.
A list to a few of the popular ones:

- [IntelliJ Inspections](https://www.jetbrains.com/help/idea/code-inspection.html)
- [PMD](https://pmd.github.io/)
- [SonarQube](https://www.sonarqube.org/)

**Examples**:

- IntelliJ IDEA introduced an inspection rule for Java 17 or higher to detect
  redundant `strictfp` modifiers. This resulted in the analysis of `RedundantModifierCheck`
  to verify if we needed to update it.

### Real Usage Examples in Large Projects

Review real usage examples of the new language feature in large projects.
This helps to identify potential issues and ensures that the checks
are aligned with practical use cases.
A list of representative projects can be found

> [here](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/github-action-projects1.properties)

Reviewer: ```suggestion
[here](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/github-action-projects1.properties).
```

  to suggest the use of this new feature if we see the typical `if X instanceof Y`,
  then typecasting pattern.

### Discover Similar Checks

Look for existing checks that address concepts
similar to those introduced by the new language feature.
These checks serve as references for designing and implementing new checks.

**Examples**:

- **Identifiers and Naming Conventions**: If the new feature introduces
  new tokens involving identifiers, explore existing checks related to naming conventions.
  We can create a new similar check to enforce naming conventions for the new feature.

## How to Create Tracker Issue

The tracker issue should document and share the results of the analysis
conducted according to the outlined procedures in this document.
It should include sections for each step of the analysis with the result
of all of our actions. These tracker issues should be created for each
new language feature, demonstrating our due diligence
in integrating the feature into Checkstyle.

### Document Discovery Methods

After thoroughly researching the impact of the new language feature using the previously
mentioned methods, document all relevant findings including impacted checks,

> possible static analysis coverage (new checks), and best practices.

Reviewer: I don't remember how we ended up with this section. I think we can remove it after seeing this work in it's (current) entirety. We should capture everything we need in the tracker ticket, this seems "extra"

Reviewer: @rnveach thoughts?

Reviewer 2: I am ok to remove it. I still think this document lacks a lot of specifics, which is blocking me, but I understand this is just a draft. Once we create our first tracker issue and sure its fluffed up, that we will know exactly what should be here and lay out all the specifics.

## How to Open Child Issues

> Open child issues for each check that needs to be updated or for a new check related

Reviewer: ```suggestion
Using your findings from the tracker issue, open child issues for each check that needs to be updated or for a new check related
```

Let's tie this in nicely here.

Reviewer: Now we have a clear picture of the work to be done.

For each new language feature:
 - read this doc
 - follow procedure
 - capture results of analysis in tracker ticket
 - create individual check tickets from tracker ticket findings
 - send PRs to close tickets
 - close tracker once all children are done
 - be done

Reviewer 2: > capture results of analysis in tracker ticket
> create individual check tickets from tracker ticket findings

https://github.com/checkstyle/checkstyle/pull/14904#discussion_r1623256646
In between here should be admins review tracker and obtain approval.

Reviewer: updated my comment

# New Language Feature Check Integration Process [WIP]

## Check Update Procedure

### Consider Similar Tokens
### Java Enhancement Proposals (JEPS)
### Always Impacted Checks

- illegalToken
- illegalTokenText 
- whitespace

> - indentation

Reviewer: Use the correct casing that is in our docs, and hyperlink it, like this: [AnnotationLocation](https://checkstyle.sourceforge.io/checks/annotation/annotationlocation.html#AnnotationLocation)

### Real Usages Examples in Large Codebases

> ### Review Other Static Analysis Tools

Reviewer: Not exactly for this line, but let's leave out the openjdk mailing list for now, and create a separate issue for this.

Reviewer: @mahfouz72 ping, please open an issue for this

Author: https://github.com/checkstyle/checkstyle/issues/14922

> # Check Update Procedure

Reviewer 2: IMO `Check Update Procedure` => `Check Update Procedure for New Language Features` or something similar.

You don't want to confuse people that this is for normal check updates like adding a new property. This is specifically if we change ANTLR. Even though the file name says the same thing, it just helps to be explicit in what the user is reading. Some will gloss over a file name, especially if it is long or hard to read.

> # Check Update Procedure

Reviewer: Let's use what is in the issue: `New Language Feature Check Integration Process [WIP]` with the clear `[WIP]` tag at the end.

> ## 1. Check Discovery Methods

Reviewer: ```suggestion
## Check Update Procedure
```

> ### 1.1 Consider Similar Tokens

Reviewer: Let's remove all numbers and let the `###` determine structure for now.

### 1.2 Java Enhancement Proposals (JEPS)

> ### 1.3 Always Impacted Checks

Reviewer: Please add sub headers here, for whitespace, illegaltoken, illegaltokentext

Author: As bullet points ok? And what about indentation should we include it as always impacted

Reviewer: We can do bullet points, this is a good idea

Reviewer: Yes, let's count indentation in whitespace for sure

Author: done

> ### 1.4 Real Usages Examples in Large codebases

Reviewer: Let's make capitalization consistent, please use "title case" for headers, for all headers.

### 1.5 Review Other static analysis tools

---

## 2. How to Create Tracker Issue

### 2.1 Document Discovery Methods
### 2.2 Share Results

---
## 3. How to Open Child Issues

___

> # New Check Procedure

Reviewer: Let's move this up to directly below `Check Update Procedure`

> ## 1. Good Source of best Practices

Reviewer: ```suggestion
## Review Established Best Practices
```

> ---

Reviewer: Let's remove all of these for now

> ## 2. JEPs Recommendations

Reviewer: ```suggestion
## Review Associated JEP's Recommendations
```

# New Language Feature Check Integration Process [WIP]

## Check Update Procedure

### Consider Similar Tokens

> it's essential to analyze existing language constructs 

Reviewer: Any developer docs should be considered "technical writing"; to this end, we should be very concise and direct. We have no need for phrasing like "It's essential". Everything is essential in good tech writing :)

Reviewer: We need to make sure we have good capitalization and grammar, don't be afraid to run your stuff through ChatGPT 😂

that share similarities with the new language feature. 
Once we've identified them, we will determine which checks support these features. 
these checks are most likely to require updates.

> for `RECORD_DEF`, it was reasonable to look at all checks 

Reviewer: For examples, let's do a bold header like this for all:

### Section

Blah blah....

**Example**

My awesome example stuff...

This looks like:
```md
### Section

Blah blah....

**Example**

My awesome example stuff...
```

Reviewer: You can also draw inspiration from here for formatting: https://github.com/checkstyle/checkstyle/pull/14514/files (make sure to check out the file view to see it rendered in all it's beauty)

Author: @nrmancuso these are great docs I can complete them later if you don't have time

Author: I remember that I took a look on them from your PR while applying and it helps a lot

that had `CLASS_DEF` in their acceptable tokens.
### Java Enhancement Proposals (JEPS)
Review Java Enhancement Proposals (JEPS) related to the new feature. 
JEPS provide detailed information about the goals
and motivations behind new language features. 
This information can help us understand the feature better and figure out
which checks are most likely to be impacted.

**For example,** Unnamed variables (`_`) the goal of this feature according to JEP 
is to declare a variable that is not intend to be used.
JEP guides us to recognize the new role of the underscore for unnamed variables

> and avoid flagging the non-use of such variables in modern code.

Reviewer: Spell it out here: we need to update BlahBlahNamingCheck to allow this pattern

### Always Impacted Checks
Certain checks are most likely to be impacted by new language features.

> Ensure these checks are updated to handle the new feature:

Reviewer: It is mandatory to analyze  impact on the following checks:
